apollo-router/src/layers/opentracing.rs

```diff
@@ -0,0 +1,131 @@
+use std::fmt::Display;
+use std::task::{Context, Poll};
+
+use apollo_router_core::{ConfigurableLayer, SubgraphRequest};
+use http::HeaderValue;
+use opentelemetry::trace::TraceContextExt;
+use schemars::JsonSchema;
+use serde::Deserialize;
+use tower::{BoxError, Layer, Service};
+use tracing::instrument::Instrumented;
+use tracing::{span, Instrument, Level, Span};
+use tracing_opentelemetry::OpenTelemetrySpanExt;
+
+#[derive(Clone, JsonSchema, Deserialize, Debug)]
+#[serde(rename_all = "snake_case")]
+pub enum PropagationFormat {
+    Jaeger,
+    ZipkinB3,
+}
+
+impl Display for PropagationFormat {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            PropagationFormat::Jaeger => write!(f, "jaeger"),
+            PropagationFormat::ZipkinB3 => write!(f, "zipkin_b3"),
+        }
+    }
+}
+
+#[derive(Clone, JsonSchema, Deserialize, Debug)]
+pub struct OpenTracingConfig {
+    format: PropagationFormat,
+}
+
+#[derive(Debug)]
+pub struct OpenTracingLayer {
+    format: PropagationFormat,
+}
+
+impl ConfigurableLayer for OpenTracingLayer {
+    type Config = OpenTracingConfig;
+    fn new(configuration: Self::Config) -> Result<Self, BoxError> {
+        Ok(Self {
+            format: configuration.format,
+        })
+    }
+}
+
+impl<S> Layer<S> for OpenTracingLayer {
+    type Service = OpenTracingService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        OpenTracingService {
+            inner,
+            format: self.format.clone(),
+        }
+    }
+}
+
+pub struct OpenTracingService<S> {
+    inner: S,
+    format: PropagationFormat,
+}
+
+impl<S> Service<SubgraphRequest> for OpenTracingService<S>
+where
+    S: Service<SubgraphRequest>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = Instrumented<<S as tower::Service<SubgraphRequest>>::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: SubgraphRequest) -> Self::Future {
+        let current_span = Span::current();
+        let span_context = current_span.context();
+        let span_ref = span_context.span();
+        let current_span_ctx = span_ref.span_context();
+        let (trace_id, parent_span_id, trace_flags) = (
+            current_span_ctx.trace_id(),
+            current_span_ctx.span_id(),
+            current_span_ctx.trace_flags(),
+        );
+
+        let new_span = span!(parent: current_span, Level::INFO, "subgraph_request");
+        let new_span_context = new_span.context();
+        let new_span_ref = new_span_context.span();
+        let span_id = new_span_ref.span_context().span_id();
```

Comment:
Just making sure I understand this. Are we are creating a new span here because we don't have an existing span that we can leverage? Also: INFO level seems high. This seems like maybe it should be DEBUG or TRACE level. Probably TRACE...

Reply:
To be spec compliant with open tracing they need 1 parent span id which is the current one for us and a new span id to have trace propagation via the headers

Reply:
Ok. I'm still not understanding why our current + parent aren't enough, but maybe we can discuss this in tomorrow's huddle? (I'm fairly sure our current span has a parent span)

Reply:
`Tracing Client Calls
When an application acts as an RPC client, it is expected to start a new tracing Span before making an outgoing request, and propagate the new Span along with that request. The following example shows how it can be done for an HTTP request.`
copy pasted from https://opentracing.io/docs/best-practices/